File: typescript/base-hmi-types/src/converters/automation-xml/HmiScriptToAutomationXmlConverter.ts
import { HmiScript } from "../../scripts/HmiScript.js";
import { HmiScriptLanguage } from "../../screens/base/HmiProperty.js";
import { HmiAutomationXmlConverter } from "./HmiAutomationXmlConverter.js";
import { XmlWriter } from "./XmlWriter.js";

export class HmiScriptToAutomationXmlConverter {
  static writeAutomationXml(writer: XmlWriter, script: HmiScript, id: { value: number }): void {
    if (script.language !== HmiScriptLanguage.VBScript) {
      return;
    }

    writer.writeStartElement("Hmi.RuntimeScripting.VBScript");
    writer.writeAttributeString("ID", String(id.value++));
    writer.writeStartElement("AttributeList");
    writer.writeElementString("Code", script.sourceCode);
    writer.writeElementString("Name", script.name);
    writer.writeElementString("PreCode", script.preCode);
    writer.writeElementString("Type", script.scriptType);
    writer.writeEndElement("AttributeList");

    if (script.comment !== undefined) {
      writer.writeStartElement("ObjectList");
      HmiAutomationXmlConverter.writeMultilingualText(script.comment, writer, id, "Comment");
      writer.writeEndElement("ObjectList");
    }

    writer.writeEndElement("Hmi.RuntimeScripting.VBScript");
  }
}

File: src/BaseHmiTypes/Converters/AutomationXml/HmiScriptToAutomationXmlConverter.cs
using System.Globalization;
using System.Xml;
using BaseHmiTypes.Screens.Base;
using BaseHmiTypes.Scripts;

namespace BaseHmiTypes.Converters.AutomationXml;

public static class HmiScriptToAutomationXmlConverter
{
    public static void WriteAutomationXml(XmlWriter writer, HmiScript script, ref int id)
    {
        if (script.Language != HmiScriptLanguage.VBScript)
            return;

        writer.WriteStartElement("Hmi.RuntimeScripting.VBScript");
        writer.WriteAttributeString("ID", (id++).ToString(CultureInfo.InvariantCulture));
        writer.WriteStartElement("AttributeList");
        WriteElementString(writer, "Code", script.SourceCode);
        WriteElementString(writer, "Name", script.Name);
        WriteElementString(writer, "PreCode", script.PreCode);
        WriteElementString(writer, "Type", script.ScriptType.ToString());
        writer.WriteEndElement();

        if (script.Comment != null)
        {
            writer.WriteStartElement("ObjectList");
            HmiAutomationXmlConverter.WriteAutomationXml(script.Comment, writer, ref id, "Comment");
            writer.WriteEndElement();
        }

        writer.WriteEndElement();
    }

    private static void WriteElementString(XmlWriter writer, string name, object? value)
    {
        writer.WriteElementString(name, value switch
        {
            null => string.Empty,
            bool boolean => boolean ? "true" : "false",
            IFormattable formattable => formattable.ToString(null, CultureInfo.InvariantCulture),
            _ => value.ToString()
        });
    }
}
